classes/local/store/s3/client.php

```diff
@@ -79,44 +79,91 @@ public function __wakeup() {
-        $this->client->registerStreamWrapper();
+        if ($this->is_functional()) {
+            $this->client->registerStreamWrapper();
+        }
     }
 
-    public function set_client($config) {
-        $options = array(
-            'region' => $config->s3_region,
-            'version' => AWS_API_VERSION
-        );
+    /**
+     * Check if the client is functional.
+     * @return bool
+     */
+    private function is_functional() {
+        return !is_null($this->client);
+    }
 
-        if (empty($config->s3_usesdkcreds)) {
-            $options['credentials'] = array('key' => $config->s3_key, 'secret' => $config->s3_secret);
+    /**
+     * Check if the client configured properly.
+     *
+     * @param \stdClass $config Client config.
+     * @return bool
+     */
+    private function is_configured($config) {
+        if (empty($config->s3_bucket)) {
+            return false;
         }
 
-        if ($config->useproxy) {
-            $options['http'] = array('proxy' => $this->get_proxy_string());
+        if (empty($config->s3_region)) {
+            return false;
         }
 
-        // Support base_url config for aws api compatible endpoints.
-        if ($config->s3_base_url) {
-            $options['endpoint'] = $config->s3_base_url;
+        if (empty($config->s3_usesdkcreds)) {
+            if (empty($config->s3_key) && empty($config->s3_secret)) {
+                return false;
+            }
         }
 
-        $this->client = \Aws\S3\S3Client::factory($options);
+        return true;
+    }
+
+    /**
+     * Set the client.
+     *
+     * @param \stdClass $config Client config.
+     */
+    public function set_client($config) {
+        if (!$this->is_configured($config)) {
+            $this->client = null;
+        } else {
+            $options = array(
+                'region' => $config->s3_region,
+                'version' => AWS_API_VERSION
+            );
+
+            if (empty($config->s3_usesdkcreds)) {
+                $options['credentials'] = array('key' => $config->s3_key, 'secret' => $config->s3_secret);
+            }
+
+            if ($config->useproxy) {
+                $options['http'] = array('proxy' => $this->get_proxy_string());
+            }
+
+            // Support base_url config for aws api compatible endpoints.
+            if ($config->s3_base_url) {
+                $options['endpoint'] = $config->s3_base_url;
+            }
+
+            $this->client = \Aws\S3\S3Client::factory($options);
+        }
     }
 
     /**
      * Registers 's3://bucket' as a prefix for file actions.
      *
      */
     public function register_stream_wrapper() {
-        if ($this->get_availability()) {
+        if ($this->get_availability() && $this->is_functional()) {
             $this->client->registerStreamWrapper();
         } else {
             parent::register_stream_wrapper();
         }
     }
 
     private function get_md5_from_hash($contenthash) {
```

Comment:
Can you please add phpdocs to this function (In particular params and return type), I see false is already expected e.g. from an exception throw but just to be clearer upfront.

Reply:
should be in different MR